lib/tree.rb

```diff
@@ -16,49 +16,133 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  # Time Complexity: O(log n)
+  # Space Complexity: ??
+  def add(key, value = nil)
+    new_node = TreeNode.new(key, value)
+
+    if @root.nil?
+      @root = new_node
+    else
+      add_helper(@root, new_node)
+    end
+  end
+
+  def add_helper(current, new_node)
+    return new_node if current.nil?
+
+    if new_node.key == current.key
+      current = new_node
+    elsif new_node.key < current.key
+      current.left = add_helper(current.left, new_node)
+    elsif new_node.key > current.key
+      current.right = add_helper(current.right, new_node)
+    end
+    return current
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n)
+  # Space Complexity: ??
   def find(key)
-    raise NotImplementedError
+    current = @root
+    current.nil? ? return : find_helper(current, key)
+  end
+
+  def find_helper(current, key)
+    if key < current.key
+      find_helper(current.left, key)
+    elsif key > current.key
+      find_helper(current.right, key)
+    else
+    return current.value
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n) Where n is the number of nodes in the tree. The algorithm visits each node exactly once
+  # And due to the worst case when the tree becomes unbalanced
+  # Space Complexity: O(n)
   def inorder
-    raise NotImplementedError
+    values = []
+    return inorder_helper(@root, values)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(current, values)
+    return values if current.nil?
+
+    inorder_helper(current.left, values)
+    values.push({ key: current.key, value: current.value })
+    inorder_helper(current.right, values)
+
+    return values
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    values = []
+    return preorder_helper(@root, values)
+  end
+
+  def preorder_helper(current, values)
+    return values if current.nil?
+
+    values.push({ key: current.key, value: current.value })
+    preorder_helper(current.left, values)
+    preorder_helper(current.right, values)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def postorder
-    raise NotImplementedError
+    values = []
+    return postorder_helper(@root, values)
+  end
+
+  def postorder_helper(current, values)
+    return values if current.nil?
+
+    postorder_helper(current.left, values)
+    postorder_helper(current.right, values)
+    values.push({ key: current.key, value: current.value })
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: ??
   def height
```

Comment:
👍 the space complexity is O(n) if the tree is unbalanced, and O(log n) if it is.